### PR TITLE
Product image management

### PR DESCRIPTION
**Pull request by Pikkupomo**
Product image management is now part of product edit form, as it was in original design.

Refs SHOOP-520

### CODE REVIEW

File: shoop/admin/gulp_bits/js-packages.js [hunk 1]
```diff
@@ -28,7 +28,7 @@ module.exports = {
             "js/content-blocks.js",
             "js/custom-selects.js",
             "js/remarkable-field.js",
-            "js/form-utils.js"
```
Comment (akx) [inline shoop/admin/gulp_bits/js-packages.js]: Unnecessary change :)

File: shoop/admin/modules/products/views/edit.py [hunk 2]
```diff
@@ -104,11 +105,56 @@ def form_valid(self, form):
...
+    name = "images"
+    priority = -1001
+    formset = ProductImageMediaFormSet
+
+    def form_valid(self, form):
+        if self.name in form.forms:
+            frm = form.forms[self.name]
+            frm.save()
+
+            # delete possible link
```
Comment (akx) [inline shoop/admin/modules/products/views/edit.py]: Is this necessary? `primary_image` has `on_delete=models.SET_NULL`...

File: shoop/admin/modules/products/views/edit_media.py [hunk 3]
```diff
@@ -21,6 +23,7 @@
 from shoop.utils.multilanguage_model_form import MultiLanguageModelForm
 
 
+# XXX: Will be deprecated when product edit view has other media management
```
Comment (akx) [inline shoop/admin/modules/products/views/edit_media.py]: Are these remarks necessary? We'll delete things when the issue for that rolls around anyway.

File: shoop/admin/modules/products/views/forms.py [hunk 4]
```diff
@@ -6,9 +6,17 @@
 # This source code is licensed under the AGPLv3 license found in the
 # LICENSE file in the root directory of this source tree.
 from __future__ import unicode_literals
+
 from django import forms
+from django.core.exceptions import ValidationError
+from django.forms import BaseModelFormSet
+from django.forms.formsets import DEFAULT_MIN_NUM
```
Comment (akx) [inline shoop/admin/modules/products/views/forms.py]: Consolidate these two imports from the same module onto a single line.

File: shoop/admin/modules/products/views/forms.py [hunk 5]
```diff
@@ -6,9 +6,17 @@
...
+from django.forms.formsets import DEFAULT_MAX_NUM
+from django.utils.translation import ugettext_lazy as _
 
-from shoop.core.models import Product, ShopProduct
+from filer.models import Image
+from shoop.admin.forms.widgets import MediaChoiceWidget
+from shoop.core.models import Product, ShopProduct, ProductMedia, ProductMediaKind, Shop
 from shoop.core.models.attributes import AttributeType, Attribute
```
Comment (akx) [inline shoop/admin/modules/products/views/forms.py]: Consolidate this import line onto the previous one. (The models are available via `.models` as well as in the submodules.)

File: shoop/admin/modules/products/views/forms.py [hunk 6]
```diff
@@ -134,3 +142,115 @@ def save(self):
...
+            "external_url",
+            "public",
+            "title",
+            "description",
+            "purchased",
+            "shops",
+            "kind"
+        )
+
+    def __init__(self, **kwargs):
```
Comment (akx) [inline shoop/admin/modules/products/views/forms.py]: Add a docstring describing `allow_only`? It's not very intuitive unless you know.

File: shoop/admin/modules/products/views/forms.py [hunk 7]
```diff
@@ -134,3 +142,115 @@ def save(self):
...
+        self.product = kwargs.pop("product")
+        self.allow_only = kwargs.pop("allow_only")
+        super(ProductMediaForm, self).__init__(**kwargs)
+
+        self.fields["file"].widget = MediaChoiceWidget()  # Filer misimplemented the field; we need to do this manually.
+        self.fields["file"].required = True
+
+        if not self.instance.pk:
+            self.fields["shops"].initial = [Shop.objects.first().pk]
```
Comment (akx) [inline shoop/admin/modules/products/views/forms.py]: This will cause N database lookups (where N = the amount of `extra`s).

File: shoop/admin/modules/products/views/forms.py [hunk 8]
```diff
@@ -134,3 +142,115 @@ def save(self):
...
+        if self.allow_only:
+            qs = qs.filter(kind=self.allow_only)
+        else:
+            qs = qs.exclude(kind=ProductMediaKind.IMAGE)
+        return qs
+
+    def form(self, **kwargs):
+        kwargs.setdefault("languages", self.languages)
+        kwargs.setdefault("product", self.product)
+        kwargs.setdefault("allow_only", self.allow_only)
```
Comment (akx) [inline shoop/admin/modules/products/views/forms.py]: `.allow_only` is not an attribute of this class.

File: shoop/admin/modules/products/views/forms.py [hunk 9]
```diff
@@ -134,3 +142,115 @@ def save(self):
...
+
+    def clean_file(self):
+        file = self.cleaned_data.get("file")
+        if file and self.allow_only == ProductMediaKind.IMAGE and not isinstance(file, Image):
+            raise ValidationError(_("Only images allowed in this field"))
+        return file
+
+    def save(self, commit=True):
+        instance = super(ProductImageMediaForm, self).save(commit)
+        print(self.cleaned_data)
```
Comment (akx) [inline shoop/admin/modules/products/views/forms.py]: Stray print.

File: shoop/admin/modules/products/views/forms.py [hunk 10]
```diff
@@ -134,3 +142,115 @@ def save(self):
...
+        return instance
+
+
+class ProductImageMediaFormSet(ProductMediaFormSet):
+    allow_only = ProductMediaKind.IMAGE
+    form_class = ProductImageMediaForm
+
+    def add_fields(self, form, index):
+        super(ProductImageMediaFormSet, self).add_fields(form, index)
+        form.fields["is_primary"] = forms.BooleanField(required=False)
```
Comment (akx) [inline shoop/admin/modules/products/views/forms.py]: You can just add `is_primary = forms.BooleanField(required=False)` to `ProductImageMediaForm` instead of doing it here.

File: shoop/admin/modules/products/views/forms.py [hunk 11]
```diff
@@ -134,3 +142,115 @@ def save(self):
...
+        if form.instance and self.product.primary_image_id == form.instance.pk:
```
Comment (akx) [inline shoop/admin/modules/products/views/forms.py]: Likewise this looks like something that should be done in the form's init instead of hijacking `add_fields`.

File: shoop/admin/static_src/product/edit_media.js [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+        if(confirm("Are you sure?"))
+        {
+            $(this).parents(".panel").fadeOut();
+            $(this).next(".hide").find("input").attr("checked", true);
+        }
+    });
+
+    $(document).on("click", ".set-as-primary", function(e){
+        e.preventDefault();
+        var $_parent = $(this).parents(".panel");
```
Comment (akx) [inline shoop/admin/static_src/product/edit_media.js]: Why `$_parent` instead of, say, `$panel`?

File: shoop/admin/static_src/product/edit_media.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+        var prefix = $_parent.data("prefix");
+        var pieces = prefix.split("-");
+
+        var start = pieces[0];
+        var current = pieces[1];
+
+        $(".panel").removeClass("panel-selected").addClass("panel-default");
```
Comment (akx) [inline shoop/admin/static_src/product/edit_media.js]: All panels on the entire page?

File: shoop/admin/static_src/product/edit_media.js [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+        $(".is-primary-image").replaceWith(function(){
+            return $("<a>", {"class": "set-as-primary", "href": "javascript:void(0);"}).text("Set as primary image");
+        });
+
+        for (var i = 0; i < $(".panel").length; i++)
```
Comment (akx) [inline shoop/admin/static_src/product/edit_media.js]: All panels on the entire page?
Also, `$(".panel").each(function(i) { ... })`

File: shoop/admin/static_src/product/edit_media.js [hunk 15]
```diff
@@ -0,0 +1,56 @@
...
+        $(this).replaceWith(function(){
+            return $("<span>", {"class": "is-primary-image"}).text("Primary image");
+        });
+        $_parent.addClass("panel-selected");
+        $("#id_images-"+current+"-is_primary").attr("checked", true);
+    });
+
+    $(".media-add-new-panel").on("click", function(e){
+        e.preventDefault();
+        var panelCount = $(".panel").length;
```
Comment (akx) [inline shoop/admin/static_src/product/edit_media.js]: Again, all panels?

File: shoop/admin/templates/shoop/admin/products/_edit_media_form.jinja [hunk 16]
```diff
@@ -0,0 +1,96 @@
+{% from "shoop/admin/macros/general.jinja" import content_block %}
+{% from "shoop/admin/macros/multilanguage.jinja" import language_dependent_content_tabs %}
+{% set media_form = form[form_def.name] %}
+
+{% macro render_media_form(f, media_form, idx) %}
+    {% for h in f.hidden_fields() %}
+        {{ h|safe }}
+    {% endfor %}
+    {%- set is_image_form = media_form.prefix == "images" %}
```
Comment (akx) [inline shoop/admin/templates/shoop/admin/products/_edit_media_form.jinja]: You assign this value from `media_form.prefix == "images"`, then only use it in one place, where the `media_form.prefix == "images"` expression is used like 5 times.

**Comment by akx** [comment]
You should add the author email in these commits to your [Github Emails](https://github.com/settings/emails).

File: shoop/admin/modules/products/views/edit.py [hunk 17]
```diff
@@ -104,11 +105,56 @@ def form_valid(self, form):
...
+
+class ProductMediaFormPart(BaseProductMediaFormPart):
+    name = "media"
+    priority = -700
+    formset = ProductMediaFormSet
+
+    def form_valid(self, form):
+        if self.name in form.forms:
+            frm = form.forms[self.name]
+            instances = frm.save()
```
Comment (akx) [inline shoop/admin/modules/products/views/edit.py]: This will end up saving the objects twice, which seems like a waste...

File: shoop/admin/modules/products/views/edit.py (hunk 2 above)
Comment (Pikkupomo) [inline shoop/admin/modules/products/views/edit.py]: For some reason, this (`on_delete`) didn't work when I first prototyped the feature. It does now, so I will remove these lines.

**Comment by akx** [comment]
Fix the unnecessary import flake8 is complaining about, please:

> shoop/admin/modules/products/views/edit.py:21:1: F401 'ProductMediaKind' imported but unused

**Comment by akx** [comment]
retest this please

File: shoop/admin/modules/products/views/forms.py [hunk 18]
```diff
@@ -134,3 +140,116 @@ def save(self):
...
+                    'upscale': True,
+                })
+                self.file_url = self.instance.url
+                self.thumbnail = thumbnail.url or None
+            if self.product.primary_image_id == self.instance.pk:
+                self.fields["is_primary"].initial = True
+
+    def clean_file(self):
+        file = self.cleaned_data.get("file")
+        if file and self.allow_only == ProductMediaKind.IMAGE and not isinstance(file, Image):
```
Comment (akx) [inline shoop/admin/modules/products/views/forms.py]: `allow_only` here.

File: shoop/admin/static_src/product/edit_media.js [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+    $(document).on("click", ".set-as-primary", function(e){
+        e.preventDefault();
+        var $panel = $(this).parents(".panel");
+        var prefix = $panel.data("prefix");
+        var pieces = prefix.split("-");
+
+        var start = pieces[0];
+        var current = pieces[1];
+
+        var image_panels = $("#product-images-section .panel");
```
Comment (akx) [inline shoop/admin/static_src/product/edit_media.js]: Use camelCase in JavaScript, and prepend jQuery variables with dollars.

File: shoop/admin/static_src/product/edit_media.js [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+ * This source code is licensed under the AGPLv3 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+$(function() {
+    $(".product-media-delete").on("click", function (e) {
+        e.preventDefault();
+        if(confirm("Are you sure?"))
+        {
+            $(this).parents(".panel").fadeOut();
+            $(this).next(".hide").find("input").attr("checked", true);
```
Comment (akx) [inline shoop/admin/static_src/product/edit_media.js]: This looks pretty brittle. What does it do?

File: shoop/admin/templates/shoop/admin/products/_edit_media_form.jinja [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+            </div>
+        </div>
+    </div>
+{% endmacro %}
+
+
+{%- set is_image_form = media_form.prefix == "images" %}
+{%- set name = "Product Images" if is_image_form else "Product Media" %}
+
+{% call content_block(_(name), "fa-image") %}
```
Comment (akx) [inline shoop/admin/templates/shoop/admin/products/_edit_media_form.jinja]: Do the translation in the `{% set %}` statement; this won't be picked up by the i18n stack.

File: shoop/admin/static_src/product/edit_media.js [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    });
+
+    $(document).on("click", ".set-as-primary", function(e){
+        e.preventDefault();
+        var $panel = $(this).parents(".panel");
+        var prefix = $panel.data("prefix");
+        var pieces = prefix.split("-");
```
Comment (akx) [inline shoop/admin/static_src/product/edit_media.js]: Enable `"es6": true` in the Gulp configuration for this file, and you can use destructuring:

```javascript
var [start, current] = prefix.split("-");
```

File: shoop/admin/modules/products/views/edit.py [hunk 23]
```diff
@@ -102,11 +105,48 @@ def form_valid(self, form):
...
+            kwargs={"product": self.object, "languages": settings.LANGUAGES}
+        )
+
+
+class ProductMediaFormPart(BaseProductMediaFormPart):
+    name = "media"
+    priority = -700
+    formset = ProductMediaFormSet
+
+    def form_valid(self, form):
```
Comment (akx) [inline shoop/admin/modules/products/views/edit.py]: Can you hoist this up to `BaseProductMediaFormPart`? Seems to be the same code in `ProductMediaFormPart` and `ProductImageMediaFormPart`.

File: shoop/admin/templates/shoop/admin/products/_edit_media_form.jinja [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+    {{ media_form.management_form }}
+    {% for f in media_form %}
+        {{ render_media_form(f, media_form, loop.index, is_image_form) }}
+    {% endfor %}
+    <div class="hide" id="placeholder-panel">
+        {{ render_media_form(media_form.empty_form, media_form, "__prefix_name__", is_image_form) }}
+    </div>
+    <hr>
+<a class="btn btn-lg btn-text media-add-new-panel" href="#">+ {% trans %}Add new image{% endtrans %}</a>
```
Comment (akx) [inline shoop/admin/templates/shoop/admin/products/_edit_media_form.jinja]: Should this be `Add new media` in some cases?

Comment (Pikkupomo) [inline shoop/admin/templates/shoop/admin/products/_edit_media_form.jinja]: This ticket is only about images, but yes, to keep the consistency so to speak, I'll change this too :)

File: shoop/admin/static_src/product/edit_media.js [hunk 25]
```diff
@@ -0,0 +1,56 @@
...
+ * This source code is licensed under the AGPLv3 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+$(function() {
+    $(".product-media-delete").on("click", function (e) {
+        e.preventDefault();
+        if(confirm("Are you sure you want to delete this media?"))
+        {
+            $(this).parents(".panel").fadeOut();
+            $(this).next(".hide").find("input").attr("checked", true);
```
Comment (akx) [inline shoop/admin/static_src/product/edit_media.js]: This still looks brittle...

**Comment by akx** [comment]
Let's roll with this. Thanks!
